### PR TITLE
[C10D] Add functional collective reduce_scatter_into_tensor_coalesced.

### PR DESCRIPTION
**Pull request by kumpera**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103086
* __->__ #101023
* #98643

Implementation uses a fallback that does no coalescing.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @soumith @desertfire

### CODE REVIEW

File: test/distributed/test_functional_api.py [hunk 1]
```diff
@@ -252,6 +252,16 @@ def test_all_gather_into_tensor_coalesced(self, device):
         self.assertEqual(torch.ones([4 * dist.get_world_size()], device=device), res[0])
         self.assertEqual(torch.ones([4 * dist.get_world_size()], device=device) + 1, res[1])
 
+    @skip_if_lt_x_gpu(1)
+    def test_reduce_scatter_into_tensor_coalesced_cuda(self):
+        tensors = [torch.ones([4], dtype=torch.int64, device="cuda:0"), torch.ones([4], dtype=torch.int64, device="cuda:0") + 1]
```
Comment (wanchaol) [inline test/distributed/test_functional_api.py:257]: nit: device should be "cuda:rank" instead of 0? if we have more than 2 GPUs on the CI this would crash

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -4999,3 +4999,47 @@ def codegen_collective(self, wrapper, output_name, input_names):
...
+        inputs = [cls.realize_input(x) for x in inputs]
+
+        def compute_size(new_size):
+            new_size[0] //= group_size
+
+        outputs = cls.create_output_buffers(inputs, compute_size)
+
+        layout = MultiOutputLayout(inputs[0].get_device())
+
+        _ = ReduceScatterTensorCoalesced(
```
[the remark this answers is not in the document]
Comment (kumpera) [reply torch/_inductor/ir.py:5063]: We have to return the nodes that map to the output of the op and the collective nodes don't represent that, the output nodes do.

This works because we mark the collective as having side effects so Inductor's scheduler will not DCE it.

**Review by wanchaol — APPROVED** [review]
lgtm, thanks!
